### PR TITLE
LibJS: Implement tagged template literals (foo`bar`)

### PR DESCRIPTION
**Pull request by linusg**
To make processing tagged template literals easier, template literals will now add one empty `StringLiteral` before and after each template expression *if* there's no other string - e.g.:

```
`${foo}` -> "", foo, ""
`test${foo}${bar}test` -> "test", foo, "", bar, "test"
```

This also matches the behaviour of many other parsers.

### CODE REVIEW

File: Libraries/LibJS/AST.cpp [hunk 1]
```diff
@@ -1297,6 +1296,46 @@ Value TemplateLiteral::execute(Interpreter& interpreter) const
...
+    m_template_literal->dump(indent + 2);
+}
+
+Value TaggedTemplateLiteral::execute(Interpreter& interpreter) const
+{
+    auto tag = m_tag->execute(interpreter);
+    if (interpreter.exception())
+        return {};
+    // FIXME: We really should have a helper for this.
+    if (!tag.is_object() || !tag.as_object().is_function()) {
```
Comment (awesomekling) [inline Libraries/LibJS/AST.cpp:1316]: How about we just add `Value::is_function()`? :)

File: Libraries/LibJS/AST.cpp [hunk 2]
```diff
@@ -1297,6 +1296,46 @@ Value TemplateLiteral::execute(Interpreter& interpreter) const
...
+        interpreter.throw_exception<TypeError>(String::format("%s is not a function", tag.to_string().characters()));
+        return {};
+    }
+    auto& tag_function = static_cast<Function&>(tag.as_object());
```
Comment (awesomekling) [inline Libraries/LibJS/AST.cpp:1320]: And `Value::as_function()`!

File: Libraries/LibJS/AST.h [hunk 3]
```diff
@@ -786,6 +786,24 @@ class TemplateLiteral final : public Expression {
     const NonnullRefPtrVector<Expression> m_expressions;
 };
 
+class TaggedTemplateLiteral final : public Expression {
+public:
+    TaggedTemplateLiteral(NonnullRefPtr<Expression> tag, NonnullRefPtr<TemplateLiteral> template_literal)
+        : m_tag(tag)
+        , m_template_literal(template_literal)
```
Comment (awesomekling) [inline Libraries/LibJS/AST.h:793]:
```suggestion
        : m_tag(move(tag))
        , m_template_literal(move(template_literal))
```

File: Libraries/LibJS/AST.cpp (hunk 1 above)
Comment (linusg) [reply Libraries/LibJS/AST.cpp]: Yes! It was about time :)

File: Libraries/LibJS/Runtime/Object.cpp [hunk 4]
```diff
@@ -477,10 +477,8 @@ Value Object::to_primitive(PreferredType preferred_type) const
 Value Object::to_string() const
 {
     auto to_string_property = get("toString");
-    if (!to_string_property.is_empty()
-        && to_string_property.is_object()
-        && to_string_property.as_object().is_function()) {
-        auto& to_string_function = static_cast<Function&>(to_string_property.as_object());
+    if (!to_string_property.is_empty() && to_string_property.is_function()) {
```
Comment (awesomekling) [inline Libraries/LibJS/Runtime/Object.cpp:480]: Even tighter:
```suggestion
    if (to_string_property.is_function()) {
```

**Pull request merged**